Lib/asyncio/streams.py

```diff
@@ -1,27 +1,121 @@
 __all__ = (
-    'StreamReader', 'StreamWriter', 'StreamReaderProtocol',
-    'open_connection', 'start_server')
+    'Stream', 'StreamMode', 'Listener',
+    'open_connection', 'start_server',
+    'connect',
+    'StreamServer')
 
+import enum
 import socket
 import sys
 import warnings
 import weakref
 
 if hasattr(socket, 'AF_UNIX'):
-    __all__ += ('open_unix_connection', 'start_unix_server')
+    __all__ += ('open_unix_connection', 'start_unix_server',
+                'connect_unix',
+                'UnixStreamServer')
 
 from . import coroutines
 from . import events
 from . import exceptions
 from . import format_helpers
 from . import protocols
 from .log import logger
-from .tasks import sleep
+from . import tasks
 
 
 _DEFAULT_LIMIT = 2 ** 16  # 64 KiB
 
 
+class StreamMode(enum.Flag):
+    READ = enum.auto()
+    WRITE = enum.auto()
+    READWRITE = READ | WRITE
+
+    def _check_read(self):
+        if not self & self.READ:
+            raise RuntimeError("The stream is write-only")
+
+    def _check_write(self):
+        if not self & self.WRITE:
+            raise RuntimeError("The stream is read-only")
+
+class Listener:
+    # The class represents listening sockets served by
+    # StreamServer and UnixStreamServer
+    # The exposed API is IPv4/IPv6/UNIX address itself
+    # and socket fileno
+    # The socket object is not exposed intentionally
+    # to prevent wild access from high-level API
+    # sock = socket.fromfd(fileno, family, type, proto)
+    # still can be used if the hacking is really necessary
+
+    __slots__ = ('_family', '_type', '_proto', '_addr', '_fileno')
+
+    def __init__(self, sock, _asyncio_internal=False):
+        if not _asyncio_internal:
+            raise TypeError(f"{self.__class__} should be instaniated "
+                            "by asyncio internals only")
+        self._family = sock.family
+        self._type = sock.type
+        self._proto = sock.proto
+        self._addr = sock.getsockname()
+        self._fileno = sock.fileno()
+
+    @property
+    def family(self):
+        return self._family
+
+    @property
+    def type(self):
+        return self._type
+
+    @property
+    def proto(self):
+        return self._proto
+
+    @property
+    def addr(self):
```

Comment:
I do want to return instances of this class from `transport.get_extra_info('socket')`.  So I'm -1 on the current design and propose to have `.getsockname()` method instead.

Reply:
The class is removed for the sake of using whatever that `asyncio.Sever.sockets` return.